### ISSUE TITLE
Order of prepends and extends can break super

### ISSUE THREAD

**Issue by headius** [jruby/jruby#4531]
In debugging the behavior of CGI::Escape (a C ext added in Ruby 2.3) I discovered problems with the superclass logic surrounding prepended and extended modules.

Specifically, the following example will blow out the stack:

```ruby
module X; end
class Foo; extend X; end
module Y; def blah; super; end; end
module X; prepend Y; end
class Foo; extend Y; end; Foo.blah
```

Small changes in order here will work properly, producing an error about "blah" not being available on Foo (for the super call).

For example, this works:

```ruby
module Y; def blah; super; end; end
module X; prepend Y; end
class Foo; extend X; end
class Foo; extend Y; end; Foo.blah
```

For the moment I will modify the order in which these operations happen in the `cgi` library to get the related tests passing. Specifically, these fail:

test/ruby/cgi/test_cgi_util.rb
```ruby
  Encoding.list.each do |enc|
    begin
      escaped = "&#39;&amp;&quot;&gt;&lt;".encode(enc)
      unescaped = "'&\"><".encode(enc)
    rescue Encoding::ConverterNotFoundError
      next
    else
      define_method("test_cgi_escapeHTML:#{enc.name}") do
        assert_equal(escaped, CGI::escapeHTML(unescaped))
      end
      define_method("test_cgi_unescapeHTML:#{enc.name}") do
        assert_equal(unescaped, CGI::unescapeHTML(escaped))
      end
    end
  end
```

Any non-ascii passed to these new optimized methods will hit the problem.

I believe this affects JRuby 9.1.8.0.

**Comment by headius** [comment]
The related tests were added to MRI in ruby/ruby@72d88bac5bbce17bf363fee0ada900c093508bcc.

**Comment by headius** [comment]
The MRI commit in question does not appear to have been backported to the 2.3 branch, but I have cherry-picked this change there anyway.

**Comment by headius** [comment]
This is still a problem in 9.2, and likely needs my work on the super_fixes branch to work properly. However there are some edges to work out in that branch and it won't be ready for 9.2. Bumping.

**Comment by headius** [comment]
The first example in the description also fails to call the method on MRI, so I'm not sure when that was valid:

```
$ cat blah.rb
module X; end
class Foo; extend X; end
module Y; def blah; super; end; end
module X; prepend Y; end
class Foo; extend Y; end; Foo.blah

$ rvm ruby-2.5.3 do ruby blah.rb
Traceback (most recent call last):
	1: from blah.rb:5:in `<main>'
blah.rb:3:in `blah': super: no superclass method `blah' for Foo:Class (NoMethodError)
```

The changes I made to cgi/util.rb also do not appear to be necessary to pass all the cgi/util tests. It appears that work in #5627 has made those tweaks unnecessary.

**Comment by headius** [comment]
This was never actually fixed, and my change in 3d8847e just disabled the CGI extension altogether.

**Comment by headius** [comment]
Actually I will open a new bug for this.

The primary example case presented above is *expected* to produce a `NoMethodError`. This bug was about the fact that we instead overflowed the call stack. That issue has indeed been fixed (perhaps by #5627 or perhaps not).

Showing old behavior and new correct error:

```
[] ~/projects/jruby $ rvm jruby-9.2.1.0 do ruby blah.rb
Error: Your application used more stack memory than the safety cap of 2048K.
Specify -J-Xss####k to increase it (#### = cap size in KB).
Specify -w for full java.lang.StackOverflowError stack trace

[] ~/projects/jruby $ ruby blah.rb
NoMethodError: super: no superclass method `blah' for Foo:Class
  method_missing at org/jruby/RubyBasicObject.java:1715
            blah at blah.rb:3
            blah at blah.rb:3
          <main> at blah.rb:5
```

There is, however, some additional problem with the CGI module that is not covered by this bug's reduced example or the fixes that resolved it. Restoring the CGI logic that combines `prepend` and `extend_object` similar to CGI causes the errors shown in the original description. Since this bug's example case **was** fixed in 9.2.7.0, I will file a new issue for the continuing problems with CGI.
